meson.build

```diff
@@ -72,32 +74,64 @@ private_cfg.set_quoted('PACKAGE_VERSION', meson.project_version())
 # Test for presence of some functions
 test_funcs = [ 'fork', 'fstatat', 'openat', 'readlinkat', 'pipe2',
                'splice', 'vmsplice', 'posix_fallocate', 'fdatasync',
-               'utimensat', 'copy_file_range', 'fallocate', 'static_assert',
-               'pthread_setname_np' ]
+               'utimensat', 'copy_file_range', 'fallocate' ]
 foreach func : test_funcs
     private_cfg.set('HAVE_' + func.to_upper(),
         cc.has_function(func, prefix: include_default, args: args_default))
 endforeach
-private_cfg.set('HAVE_SETXATTR', 
-        cc.has_function('setxattr', prefix: '#include <sys/xattr.h>'))
-private_cfg.set('HAVE_ICONV', 
-        cc.has_function('iconv', prefix: '#include <iconv.h>'))
-private_cfg.set('HAVE_BACKTRACE',
-        cc.has_function('backtrace', prefix: '#include <execinfo.h>'))
 
-# Test if headers exist
-private_cfg.set('HAVE_LINUX_CLOSE_RANGE_H',
-        cc.check_header('#include <linux/close_range.h>'))
+# Special case checks that need custom code
+special_funcs = {
+    'static_assert': '''
+        #include <assert.h>
+        static_assert(1, "test");
+        int main(void) { return 0; }
+    ''',
+    'pthread_setname_np': '''
+        #define _GNU_SOURCE
+        #include <pthread.h>
+        int main(void) {
+            pthread_t thread = pthread_self();
+            pthread_setname_np(thread, "test");
+            return 0;
+        }
+    ''',
+    'close_range': '''
+        #define _GNU_SOURCE
+        #include <unistd.h>
+        #include <fcntl.h>
+        #include <linux/close_range.h>
+        int main(void) {
+            unsigned int flags = CLOSE_RANGE_UNSHARE;
+            return close_range(3, ~0U, flags);
+        }
+    '''
+}
+
+foreach name, code : special_funcs
+    private_cfg.set('HAVE_' + name.to_upper(),
+        cc.compiles(code,
+                   args: ['-D_GNU_SOURCE', '-I/usr/include'] + args_default,
```

Comment:
This gives problems with cross-compilation, because it looks into host folder

Reply:
Indeed in Buildroot it can't find any of the 3 special functions

Reply:
'-I/usr/include' on Buildroot side is useless and I think it's useless in general because that should be a standard path of the toolchain. Am I right?

Reply:
cc.compiles() is not enough, we could go with it with -Werror or even better cc.links(), that way I've tested with NPTL and linuxthread and it recognizes correctly pthread_setname_np(). As is now it doesn't work because of:
```
- cc.compiles()
- -I/usr/include
```

Reply:
Thank you, I changed to `cc.links()`, added `-Werror` and removed `-I/usr/include`

Reply:
Ok, just FYI, -Werror was mandatory to treat the missing function prototypes as errors. But since you're using cc.links() here you can remove -Werror. We only need to make sure cc.links() is there since the beginning of meson since I've seen other projects using cc.compiles() with -Werror option enabled instead of cc.links().
Anyway it works as is, simply it returns with error when it can't find the function prototype instead of failing during linking.

Thank you for the solution provided that fixed my patch :-)

Reply:
I don't have a strong opinion about it - changed it back to cc.compiles with the -Werror. Thank a lot for your help on this!